File: backend/src/main/java/com/example/SpaResource.java
package com.example;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import java.io.InputStream;

@Path("/")
public class SpaResource {

    @GET
    @Path("/{path:.*}")
    @Produces(MediaType.TEXT_HTML)
    public Response spa(@PathParam("path") String path) {
        // Don't intercept API routes
        if (path.startsWith("api/")) {
            return Response.status(404).build();
        }
        
        // Serve index.html for all other routes (SPA routing)
        InputStream indexHtml = getClass().getResourceAsStream("/META-INF/resources/index.html");
        if (indexHtml != null) {
            return Response.ok(indexHtml).build();
        }
        return Response.status(404).build();
    }
}

File: backend/src/main/kotlin/com/example/SpaResource.kt
package com.example

import jakarta.annotation.security.PermitAll
import jakarta.annotation.security.RolesAllowed
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.PathParam
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import java.io.InputStream

@Path("/")
@PermitAll
class SpaResource {

    @GET
    @Produces(MediaType.TEXT_HTML)
    @PermitAll
    fun homepage(): Response {
        // Homepage is accessible without authentication
        val indexHtml: InputStream? = javaClass.getResourceAsStream("/META-INF/resources/index.html")
        return if (indexHtml != null) {
            Response.ok(indexHtml).build()
        } else {
            Response.status(404).build()
        }
    }

    @GET
    @Path("/introduction")
    @Produces(MediaType.TEXT_HTML)
    @PermitAll
    fun introduction(): Response {
        // Introduction page is accessible without authentication
        val indexHtml: InputStream? = javaClass.getResourceAsStream("/META-INF/resources/index.html")
        return if (indexHtml != null) {
            Response.ok(indexHtml).build()
        } else {
            Response.status(404).build()
        }
    }

    @GET
    @Path("/{path:.*}")
    @Produces(MediaType.TEXT_HTML)
    @PermitAll
    fun spa(@PathParam("path") path: String): Response {
        // Don't intercept API routes and actuator routes
        if (path.startsWith("api/") || path.startsWith("actuator/") || path.startsWith("q/") || path.startsWith("test")) {
            return Response.status(404).build()
        }
        
        // Serve index.html for all SPA routing (no authentication required)
        val indexHtml: InputStream? = javaClass.getResourceAsStream("/META-INF/resources/index.html")
        return if (indexHtml != null) {
            Response.ok(indexHtml).build()
        } else {
            Response.status(404).build()
        }
    }
}